File: LeetCode/1022. Sum of Root To Leaf Binary Numbers/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

/**
 * 63 / 63 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 16.2 MB 
 */
class Solution {
public:
    void dfs(TreeNode *node, int bit, int &ans) {
        if (node != nullptr) {
            bit = bit * 2 + node->val;
            if (node->left == nullptr && node->right == nullptr) {
                ans += bit;
            }
            dfs(node->left, bit, ans);
            dfs(node->right, bit, ans);
        }
    }

    int sumRootToLeaf(TreeNode* root) {
        int ans = 0;
        dfs(root, 0, ans);
        return ans;
    }
};

File: LeetCode/1022. Sum of Root To Leaf Binary Numbers/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
63 / 63 test cases passed.
Runtime: 40 ms
Memory Usage: 15.3 MB
"""
class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        ans = 0
        def dfs(node, bit):
            nonlocal ans
            if node:
                bit = bit * 2 + node.val
                if not node.left and not node.right:
                    ans += bit
                dfs(node.left, bit)
                dfs(node.right, bit)
        dfs(root, 0)
        return ans
